Add tutorial on symmetric tensors

leburgel opened the pull request:
A first attempt at an in-depth but (hopefully) intuitive tutorial on constructing symmetric tensors representing Hamintonians in concrete physical systems using TensorKit.jl.

leburgel committed:
Add tutorial on symmtric tensors

Jutho commented:
This looks amazing. I will try to read it in detail soon.

Jutho commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+ZZ
+```
+
+Indeed, the resulting `TensorMap` exactly encodes the matrix elements of the $ZZ$ operator shown in {ref}`the diagrams above <Z2_fusiontrees>`. The $X$ operator can be constructed way. Since it is by definition diagonal in the irrep basis with blocks directly corresponding to the trivial and sign irrep, its construction is particularly simple:
````
The $X$ operator can be constructed in a similar way. (?)

Jutho commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+:tags: [hide-output]
+a⁺a⁻ = TensorMap(zeros, ComplexF64, V ⊗ V ← V ⊗ V)
+for (s, f) in fusiontrees(a⁺a⁻)
+    if s.uncoupled[1].charge == f.uncoupled[1].charge + 1 && s.uncoupled[2].charge == f.uncoupled[2].charge - 1
```
Since it is not particularly nice that you need to access the inner `charge` field, I guess you could also write this as
`s.uncoupled[1] == only(f.uncoupled[1] ⊗ U1Irrep(1)) && s.uncoupled[2] == only(f.uncoupled[2] ⊗ U1Irrep(-1))` and similarly, in line with the example above. However, since also the matrix element depends on the `charge` field and I don't see a way around this, this is probably fine.

leburgel replied:
Checking the constraints on the uncoupled and coupled irreps using the fields is indeed not very pretty, I'll change it to your suggestion.

Since accessing the fields of a `Sector` subtype instance is basically always necessary to set the matrix elements, I think it would be good to explicitly document the field names of all sectors defined in TensorKit.jl. I'll file a PR for that soon.

Jutho commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+:tags: [hide-output]
+a⁺ = TensorMap(zeros, ComplexF64, V ← V ⊗ A)
+for (s, f) in fusiontrees(a⁺)
+    a⁺[s, f] .= sqrt(f.coupled.charge)
```
I was confused for a second here by the use of `f.coupled`, I would have thought `f.uncoupled[1].charge+1`, or `s.uncoupled[1].charge`. But both are of course equal to the coupled charge.

Jutho commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+@test norm(a⁺a⁻_bis - a⁺a⁻) ≈ 0 atol=1e-14
+@test norm(a⁻a⁺_bis - a⁻a⁺) ≈ 0 atol=1e-14
+@test norm(N_bis - N) ≈ 0 atol=1e-14
```
Would exact equality fail here? Since it is just zero and one, I would think `==` would work.

leburgel replied:
All of the 'bis' ones are built out of parts with roots in them, so they all have some floating point issues.

Jutho replied:
Right, clearly it was too late when I was reading this.

Jutho commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+```{code-cell} julia
+permute(t, ((1,), (3, 2)))
+```
+In other words, when exchanging the two codomain vector spaces, the block of the tensor map for which both corresponding irreps are odd picks up a minus sign, exactly as we would expect for fermionic charges.
````
you are exchanging the two domain spaces instead of codomain.

Jutho commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+## Level 4: Non-Abelian Symmetries and the Quantum Heisenberg Model
+
+We will now move on to systems which have more complicated *non-Abelian* symmetries. For a non-Abelian symmetry group $G$, the fact that its elements do not all commute has a profound impact on its representation theory. In particular, the irreps of such a group can be higher dimensional, and the fusion of two irreps can give rise to multiple different irreps. On the one hand this means that fusi ...
```
On the other hand, it follows (forgotten s)

Jutho commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+end
+```
+
+Based on this discussion, we can quantify the aforementioned 'difficulties' in the inverse operation of what we just demonstrated, namely converting a given operator to a symmetric tensor map given only its matrix elements in the irrep basis. Indeed, it is now clear that this precisely requires isolating the reduced matrix elements introduced above. Given the matrix elements of the operator in th ...
````
irreep -> irrep

Jutho commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+    # check that it is proportianal to the identity and return diagonal element
+    @assert norm(reduced_matrix - diagm(diag(reduced_matrix))) < 1e-12
+    @assert norm(diag(reduced_matrix) .- diag(reduced_matrix)[1]) < 1e-12
```
Isn't `norm(reduced_matrix - reduced_matrix[1,1] * I)` easier?

lkdvos replied:
I quite like just using `isapprox(reduced_matrix, reduced_matrix[1, 1] * I; atol)`

Jutho replied:
Right, indeed. No reason to put norms explicitly, this is done automatically.

Jutho reviewed (approved):
Look's great!

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+# A Symmetric Tensor Deep Dive: Constructing Your First Tensor Map
+
+In this tutorial, we will demonstrate how to construct specific tensor maps which are relevant to some common physical systems, with an increasing degree of complexity. We will assume the reader has gone through the tutorial sections on [tensor network theory](tensor_networks) and [symmetries in tensor networks](symmetric_tensors). In going through these examples we aim to provide a relatively ge ...
```
```suggestion
In this tutorial, we will demonstrate how to construct specific tensor maps which are relevant to some common physical systems, with an increasing degree of complexity. We will assume the reader has gone through the tutorial sections on [tensor network theory](tensor_networks) and [symmetries in tensor networks](symmetric_tensors). In going through these examples we aim to provide a relatively gentle introduction to the meaning of [symmetry sectors](https://jutho.github.io/TensorKit.jl/latest/man/sectors/#ss_sectors) and [vector spaces](https://jutho.github.io/TensorKit.jl/latest/man/sectors/#ss_rep) within the context of [TensorKit.jl](https://jutho.github.io/TensorKit.jl/latest/), [how to initialize a `TensorMap` over a given vector space](https://jutho.github.io/TensorKit.jl/latest/man/tensors/#ss_tensor_construction) and finally how to manually set the data of a symmetric `TensorMap`. We will keep our discussion as intuitive and simple as possible, only adding as many technical details as strictly necessary to understand each example. When considering a different physical system of interest, you should then be able to adapt these recipes and the intuition behind them to your specific problem at hand.
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+We will circle back to the implications of this symmetry later.
+
+As a warmup we will implement the Hamiltonian {eq}`eq:isingham` in the standard by encoding the matrix elements of the single-site operators $X$ and $Z$ into a complex array, and then combine them in a suitable way to get the Hamiltonian terms. Instead of using plain Julia arrays, we will use a representation in terms of `TensorMap`s over complex vector spaces. These will essentially just be wrap ...
```
```suggestion
As a warmup we will implement the Hamiltonian {eq}`eq:isingham` in the standard way by encoding the matrix elements of the single-site operators $X$ and $Z$ into an array of complex numbers, and then combine them in a suitable way to get the Hamiltonian terms. Instead of using plain Julia arrays, we will use a representation in terms of `TensorMap`s over complex vector spaces. These will essentially just be wrappers around base arrays at this point, but their construction requires some consideration of the notion of *spaces*, which generalize the notion of `size` for arrays. Each of the operators $X$ and $Z$ acts on a local 2-dimensional complex vector space. In the context of TensorKit.jl such a space can be represented as `ComplexSpace(2)`, or using the convenient shorthand `ℂ^2`. A single-site Pauli operator maps from a domain physical space to a codomain physical space, and can therefore be represented as instances of a `TensorMap(..., ℂ^2 ← ℂ^2)`. The corresponding data can then be filled in by hand according to the familiar Pauli matrices in the following way:
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+```
+
+```{note}
+In order to combine these local operators into a concrete Hamiltonian that can be using in [MPSKit.jl](https://github.com/maartenvd/MPSKit.jl) we can make use of the conventient `@mpoham` macro exported by [MPSKitModels.jl](https://github.com/maartenvd/MPSKitModels.jl). For an infinite translation invariant Ising chain, we can use the following piece of code which produces the Hamiltonian in an i ...
````
```suggestion
In order to combine these local operators into a concrete Hamiltonian that can be used in [MPSKit.jl](https://github.com/maartenvd/MPSKit.jl) we can make use of the convenient `@mpoham` macro exported by [MPSKitModels.jl](https://github.com/maartenvd/MPSKitModels.jl). For an infinite translation invariant Ising chain, we can use the following piece of code which produces the Hamiltonian in an interesting-looking form (see MPSKit.jl for details on this format).
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+:align: center
+```
+
+Recalling the [discussion on symmetries in tensor networks](symmetric_tensors), we recognize that these identitities precisely mean that these local tensors transform trivially under a tensor product representation of $\mathbb{Z}_2$. This implies that, in an appropriate basis for the local physical vector space, our local tensors would become block-diagonal where each block is labeled by a tensor ...
````
I think the blocks are labeled by just the irreps, not a tensor product of?
```suggestion
Recalling the [discussion on symmetries in tensor networks](symmetric_tensors), we recognize that these identitities precisely mean that these local tensors transform trivially under a tensor product representation of $\mathbb{Z}_2$. This implies that, in an appropriate basis for the local physical vector space, our local tensors would become block-diagonal where each block is labeled by a $\mathbb{Z}_2$ irrep. From the same discussion, we recall that the appropriate local basis transformation is precisely the one that brings the local representation $X$ into block-diagonal form. Clearly, this transformation is nothing more than the Hadamard transformation which maps the computational basis of $Z$ eigenstates $\{\ket{\uparrow}, \ket{\downarrow}\}$ to that of the $X$ eigenstates $\{\ket{+}, \ket{-}\}$ defined as $\ket{+} = \frac{\ket{\uparrow} + \ket{\downarrow}}{\sqrt{2}}$ and $\ket{-} = \frac{\ket{\uparrow} - \ket{\downarrow}}{\sqrt{2}}$. In the current context, this basis is referred to as the *irrep basis* of $\mathbb{Z}_2$, where the local basis state $\ket{+}$ corresponds to the trivial representation of $\mathbb{Z}_2$ while $\ket{-}$ corresponds to the sign representation.
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+Recalling the [discussion on symmetries in tensor networks](symmetric_tensors), we recognize that these identitities precisely mean that these local tensors transform trivially under a tensor product representation of $\mathbb{Z}_2$. This implies that, in an appropriate basis for the local physical vector space, our local tensors would become block-diagonal where each block is labeled by a tensor ...
+
+Next, let's make the statement the statement that 'the blocks of the local tensors are labeled by a tensor product of $\mathbb{Z}_2$ irreps' more concrete. To this end, consider the action of $ZZ$ in the irrep basis, which is given by the four nonzero matrix elements
```
```suggestion
Next, let's make the statement that 'the blocks of the local tensors are labeled by $\mathbb{Z}_2$ irreps' more concrete. To this end, consider the action of $ZZ$ in the irrep basis, which is given by the four nonzero matrix elements
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+\end{align}
+```
+
+If we denote the trivial $\mathbb{Z}_2$ irrep by $'0'$, corresponding to a local $\ket{+}$ state, and the sign irrep by $'1'$, corresponding to a local $\ket{-}$ state, and recall that in this notation the fusion rules of $\mathbb{Z}_2$ are given by addition modulo 2, we can associate each of the above matrix elements to a so-called *fusion tree* of $\mathbb{Z}_2$ irreps with a corresponding coëf ...
````
```suggestion
If we denote the trivial $\mathbb{Z}_2$ irrep by $'0'$, corresponding to a local $\ket{+}$ state, and the sign irrep by $'1'$, corresponding to a local $\ket{-}$ state, and recall that in this notation the fusion rules of $\mathbb{Z}_2$ are given by addition modulo 2, we can associate each of the above matrix elements to a so-called *fusion tree* of $\mathbb{Z}_2$ irreps with a corresponding coefficient of 1,
```

lkdvos replied:
I think it might be nice to distinguish the use of the word *block*. In the first part, you say that the tensors become block-diagonal, which corresponds a matrix with a label of a single (coupled) charge, and then afterwards you explain more in terms of the *small blocks*, which correspond to multidimensional arrays with a label of fusiontree / tensor product of labels

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+For our current application only `uncoupled` and `coupled` are relevant, since $\mathbb{Z}_2$ irreps are self-dual and have Abelian fusion rules. We will come back to these other properties when discussion more involved applications. Given some `TensorMap`, the method `TensorKit.fusiontrees(t::TensorMap)` returns an iterator over all pairs of splitting and fusion trees that label the blocks of `t ...
+
+We can now put this into practice by directly constructing the $ZZ$ operator in the irrep basis as a $\mathbb{Z}_2$-symmetric `TensorMap`. We will do this in two steps:
```
```suggestion
We can now put this into practice by directly constructing the $ZZ$ operator in the irrep basis as a $\mathbb{Z}_2$-symmetric `TensorMap`. We will do this in three steps:
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+ZZ = TensorMap(zeros, ComplexF64, V ⊗ V ← V ⊗ V)
+```
+
+The output of this command again demonstrates the underlying structure of a symmetric tensor. We see that all 8 valid fusion trees with 2 incoming irreps and 2 outgoing irreps {ref}`of the type above <fusiontree>` are listed with their corresponding block data. Each of these blocks is an array of shape $(1, 1, 1, 1)$ since each irrep occuring in the space $V$ has degeneracy 1. Using the `fusiontr ...
````
```suggestion
The output of this command again demonstrates the underlying structure of a symmetric tensor. We see that all eight valid fusion trees with two incoming irreps and two outgoing irreps {ref}`of the type above <fusiontree>` are listed with their corresponding block data. Each of these blocks is an array of shape $(1, 1, 1, 1)$ since each irrep occuring in the space $V$ has degeneracy 1. Using the `fusiontrees` method and the fact that we can index a `TensorMap` using a splitting/fusion tree pair, we can now fill in the nonzero blocks of the operator by observing that the $ZZ$ operator flips the irreps of the uncoupled charges in the domain with respect to the codomain, as shown in the diagrams above. Flipping a given `Z2Irrep`s in the codomain can be implemented by fusing them with the sign irrep `Z2Irrep(1)`, giving:
```
might be more clear to not use an anonymous function here, and explicitly write out flip:
```julia
flip_charge(charge::Z2Irrep) = only(charge \otimes Z2Irrep(1))
# ...
map(flip_charge, f.uncoupled)
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+For our next example we will consider the [Bose-Hubbard model](https://en.wikipedia.org/wiki/Bose%E2%80%93Hubbard_model), which describes interacting bosons on a lattice. The Hamiltonian of this model is given by
+```{math}
+:label: eq:bhh
+H = -t \sum_{\langle i,j \rangle} \left( a_{i}^+ a_{j}^- + a_{i}^- a_{j}^+ \right) - \sum_i \mu N_i + \frac{U}{2} \sum_i N_i(N_i - 1).
````
```suggestion
H = -t \sum_{\langle i,j \rangle} \left( a_{i}^+ a_{j}^- + a_{i}^- a_{j}^+ \right) - \mu \sum_i  N_i + \frac{U}{2} \sum_i N_i(N_i - 1).
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+N \ket{n} &= n \ket{n}
+\end{align}
+```
+Their bosonic nature can be summarized by familiar the commutation relations
````
```suggestion
Their bosonic nature can be summarized by the familiar commutation relations
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+```{math}
+U = \sum_i N_i
+```
+This invariance corresponds to a $\mathrm{U}(1)$ particle number symmetry, which can again be manifestly imposed when constructing the Hamiltonian terms as tensor maps. Recall that the irreps of $\mathrm{U}(1)$ are labeled by integers, and that the fusion of two irreps is given by addition.
````
```suggestion
This invariance corresponds to a $\mathrm{U}(1)$ particle number symmetry, which can again be manifestly imposed when constructing the Hamiltonian terms as `TensorMap`s. The irreps of $\mathrm{U}(1)$ are labeled by integers, and that the fusion of two irreps is given by addition.
```

lkdvos replied:
Not recall, this has not been used before I think

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+```
+
+```{note}
+Although we have made a suggestive distinction between the 'left' and 'right' versions of the operators $a_L^\pm$ and $a_R^\pm$, these can actually be represented by the same tensor map due to the [bosonic braiding style](https://jutho.github.io/TensorKit.jl/latest/lib/sectors/#TensorKit.BraidingStyle) of $\mathrm{U}(1)$ irreps. This is no longer the case when considering fermionic systems where, ...
````
Maybe *same tensormap* is also a bit suggestive, you definitely still have to permute the objects, and explicitly mentioning the permutation might also motivate the fermionic issues

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+## Level 3: Fermions and the Kitaev Model
+
+While we have already covered quite a lot of ground towards understanding symmetric tensors in terms of fusion trees and corresponding blocks, the symmetries considered so far have been quite basic in the sense that sectors corresponding to irreps of $\mathbb{Z}_2$ and $\mathrm{U}(1)$ have [*Abelian fusion rules*](https://jutho.github.io/TensorKit.jl/latest/lib/sectors/#TensorKit.FusionStyle) and ...
```
maybe not *basic* symmetries, but *simple*? sounds a bit less condescending

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+```
+This fermion parity symmetry, which we will denote as $f\mathbb{Z}_2$, is a $\mathbb{Z}_2$-like symmetry in the sense that it has a trivial representation, which we call *even* and again denote by '0', and a sign representation which we call *odd* and denote by '1'. The fusion rules of these irreps are the same as for $\mathbb{Z}_2$. Similar to the previous case, the local symmetry operator $Q_i$ ...
+
+In TensorKit.jl, an $f\mathbb{Z}_2$-graded vector spaces is represented as a `Vect[FermionParity]` space, where a given $f\mathbb{Z}(2)$ irrep can be represented as a [`FermionParity`](https://jutho.github.io/TensorKit.jl/stable/lib/sectors/#TensorKit.FermionParity) sector instance. Using the simplest instance of a vector space containing a single even and odd irrep, we can already demonstrate th ...
````
```suggestion
In TensorKit.jl, an $f\mathbb{Z}_2$-graded vector spaces is represented as a `Vect[FermionParity]` space, where a given $f\mathbb{Z}_2$ irrep can be represented as a [`FermionParity`](https://jutho.github.io/TensorKit.jl/stable/lib/sectors/#TensorKit.FermionParity) sector instance. Using the simplest instance of a vector space containing a single even and odd irrep, we can already demonstrate the corresponding fermionic braiding behavior by [performing a permutation](https://jutho.github.io/TensorKit.jl/stable/lib/tensors/#TensorKit.permute-Union{Tuple{N%E2%82%82},%20Tuple{N%E2%82%81},%20Tuple{S},%20Tuple{AbstractTensorMap{S},%20Tuple{Tuple{Vararg{Int64,%20N%E2%82%81}},%20Tuple{Vararg{Int64,%20N%E2%82%82}}}}}%20where%20{S,%20N%E2%82%81,%20N%E2%82%82}) on a simple tensor map.
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+### Constructing the Hamiltonian
+
+We can directly construct the Hamiltonian terms as symmetric tensor maps using the same procedure as before starting from their matrix elements in the occupation number basis. However, in this case we should be a bit more careful about the precise definition of the basis states in composite systems. Indeed, the tensor product structure of fermionic systems is inherently tricky to deal with, and s ...
```
```suggestion
We can directly construct the Hamiltonian terms as symmetric `TensorMap`s using the same procedure as before starting from their matrix elements in the occupation number basis. However, in this case we should be a bit more careful about the precise definition of the basis states in composite systems. Indeed, the tensor product structure of fermionic systems is inherently tricky to deal with, and should ideally be treated in the context of [*super vector spaces*](https://en.wikipedia.org/wiki/Super_vector_space). For two sites, we can define the following basis states on top of the fermionic vacuuum $\ket{00}$:
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+\end{align*}
+```
+
+Once we have these matrix elements the hard part is done, and we can naively associate these to the following $f\mathbb{Z}(2)$ fusion trees with corresponding block values,
````
```suggestion
Once we have these matrix elements the hard part is done, and we can naively associate these to the following $f\mathbb{Z}_2$ fusion trees with corresponding block values,
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+:scale: 12%
+:name: fZ2_fusiontrees
+```
+Given this information, we can go through the same procedure again to construct $c^+ c^-$, $c^- c^+$ and $N$ operators as `TensorMap`s over $f\mathbb{Z}(2)$-graded vector spaces.
````
```suggestion
Given this information, we can go through the same procedure again to construct $c^+ c^-$, $c^- c^+$ and $N$ operators as `TensorMap`s over $f\mathbb{Z}_2$-graded vector spaces.
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+### Block Sparsity Revisited: The Wigner-Eckart Theorem
+
+Let us recall some basics of representation theory first. Consider a group $G$ and a corresponding representation space $V$, such that every element $g \in G$ can be realized as a unitary operator $U_g : V \to V$. Let $h$ be a tensor map whose domain and codomain is given by the tensor product of two of these representation spaces. Recall that, by definition, the statement that '$h$ is symmetric  ...
```
```suggestion
Let us recall some basics of representation theory first. Consider a group $G$ and a corresponding representation space $V$, such that every element $g \in G$ can be realized as a unitary operator $U_g : V \to V$. Let $h$ be a `TensorMap` whose domain and codomain are given by the tensor product of two of these representation spaces. Recall that, by definition, the statement that '$h$ is symmetric under $G$' means that
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+```
+for every $g \in G$. If we label the irreducible representations of $G$ by $l$, then any representation space can be decomposed into a direct sum of irreducible representations, $V = \bigoplus_l V^{(l)}$, in such a way that $U_g$ is block-diagonal where each block is labeled by a particular irrep $l$. For each irrep space $V^{(l)}$ we can define an orthonormal basis labeled as $\ket{l, m}$, where ...
+
+[From the representation theory of groups](https://en.wikipedia.org/wiki/Tensor_product_of_representations#Clebsch%E2%80%93Gordan_theory), it is known that the product of two irreps can in turn be decomposed into a direct sum of irreps, $V^{(l_1)} \otimes V^{(l_2)} \cong \bigoplus_{l_3} V^{(l_3)}$. The precise nature of this decomposition, also refered to as the *Clebsch-Gordan problem*, is given ...
````
I think using this notation, the clebsch gordans should be the splitting  tree below, i.e. how you decompose the coupled into uncoupled basis, while the adjoint clebsch gordans are the fusion tree

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+:name: wignereckart
+```
+Here, the sum runs over all possible irreps $k$ in the fusion product $l_3 \otimes l_4$ and over all basis states $\ket{k,n}$ of $V^{(k)}$. The reduced matrix elements $h_{\text{red}}$ are independent of the basis state labels and only depend on the irrep labels themselves. Each reduced matrix element should be interpreted as corresponding to an irrep fusion tree,
+```{figure} ../_static/SymmetricTensors/anotherfusiontree.svg
````
I think the figure is a bit misleading, as typically the fusiontree drawing signifies the clebsch gordans, while this equation seems to say that these are the reduced matrix elements

leburgel replied:
True, I think I'll replace the equals sign with a `\rightarrow` in all the fusiontree-arrayblock-pair figures instead to emphasize that it's a label-value relation. And I'll add a comment that the fusion tree itself then actually implies the Clebsch-Gordans.

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+```
+The Clebsch-Gordan coefficients $C_{k}^{l_3,l_4}$ and conjugate coefficients $C^{\dagger k}_{l_1,l_2}$ encode the coupling of the domain basis states $\ket{l_3,m_3} \otimes \ket{l_4,m_4}$ to the coupled basis state $\ket{k,n}$ and the splitting of the coupled basis state to the codomain basis states $\ket{l_1,m_1} \otimes \ket{l_2,m_2}$ respectively.
+
+Since the Wigner-Eckart theorem demands that this structure in terms of Clebsch-Gordan coefficients is necessary to ensure that the corresponding is symmetric, it is precisely this structure that is inherently encoded into the fusion tree part of a symmetric tensor map. In particular, **the block value associated to each fusion tree in a symmetric tensor is precisely the reduced matrix element in ...
````
```suggestion
The Wigner-Eckart theorem demands that this structure in terms of Clebsch-Gordan coefficients is necessary to ensure that the corresponding operator is symmetric. It is precisely this structure that is inherently encoded into the fusion tree part of a symmetric tensor map. In particular, **the block value associated to each fusion tree in a symmetric tensor is precisely the reduced matrix element in the Clebsch-Gordan decomposition**.
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+Since the Wigner-Eckart theorem demands that this structure in terms of Clebsch-Gordan coefficients is necessary to ensure that the corresponding is symmetric, it is precisely this structure that is inherently encoded into the fusion tree part of a symmetric tensor map. In particular, **the block value associated to each fusion tree in a symmetric tensor is precisely the reduced matrix element in ...
+
+As a small demonstration of this fact, we can make a simple $\mathrm{SU}(2)$-symmetric tensor with trivial block values and verify that its implied symmetry structure exactly corresponds to the expected Clebsch-Gordan coefficient. In TensorKit.jl, a $\mathrm{SU}(2)$-graded vector space is represented as an [`SU2Space`](https://jutho.github.io/TensorKit.jl/latest/lib/spaces/#TensorKit.SU2Space), w ...
```
```suggestion
As a small demonstration of this fact, we can make a simple $\mathrm{SU}(2)$-symmetric tensor with trivial block values and verify that its implied symmetry structure exactly corresponds to the expected Clebsch-Gordan coefficient. In TensorKit.jl, a $\mathrm{SU}(2)$-graded vector space is represented as an [`SU2Space`](https://jutho.github.io/TensorKit.jl/latest/lib/spaces/#TensorKit.SU2Space), where a given $\mathrm{SU}(2)$ irrep can be represented as an [`SU2Irrep`](https://jutho.github.io/TensorKit.jl/latest/lib/sectors/#TensorKit.SU2Irrep) instance of integer or halfinteger spin as encoded in its `j` field. If we construct a `TensorMap` whose symmetry structure corresponds to the coupling of two spin-$\frac{1}{2}$ irreps to a spin-$1$ irrep, we can then convert it to a plain array and compare it to the $\mathrm{SU}(2)$ Clebsch-Gordan coefficients exported by the [WignerSymbols.jl package](https://github.com/Jutho/WignerSymbols.jl).
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+:name: none2symm
+```
+
+This procedure works for any group symmetry, and all we need are matrix elements of the operator in the irrep basis and the Clebsch-Gordan coefficients. In the following we demonstrate this explicit procedure for the particular example of $G = \mathrm{SU}(2)$. However, it should be noted that for general groups the Clebsch-Gordan coefficients may not be as easy to compute (if it is even known how ...
````
You could even point back to the Hadamard of the first part, and state that this is also just a Clebsch-Gordan, albeit a simple one

leburgel replied:
I'm a bit confused by this. I thought the Hadamard itself was just the basis transform from the group element to the irrep basis, I never interpreted this as having anything to do with Clebsch-Gordans. Do you mean Hadamards can be used to obtain the Clebsch-Gordans directly in the group element basis, as like a comment on the unitary freedom in their definition?

Jutho replied:
I also don't think this has anything to do with CGs. For an abelian group, all the CGs are trivial and one-dimensional, so you cannot get matrices with sqrt(2) factors.

What might be worth noting (though at most in a very small footnote) is that for abelian groups, the group elements are themselves symmetric (1,1) tensors and thus need to be diagonal. For the non-abelian case, they are not. In a (compact) Lie group, the generators are encoded in a (1,2) tensor where the extra domain leg corresponds to the adjoint representation.

lkdvos replied:
I think I got confused here, I am definitely wrong.

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+### The 'Generic' Approach to the Spin-1 Heisenberg Model: Wigner-Eckart in Action
+
+Define full array, divide out Clebsch-Gordan coefficients, separate out block value corresponding to each fusion tree.
```
This is probably still a personal note?

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+```{math}
+H = J \sum_{\langle i,j \rangle} \vec{S}_i \cdot \vec{S}_j
+```
+where $\vec{S} = (S^x, S^y, S^z)$ are the spin operators. The physical Hilbert space at each site is the three-dimensional spin-1 irrep of $\mathrm{SU}(2)$. Each two-site exchange operator $\vec{S}_i \cdot \vec{S}_j$ in the sum commutes with a global transformation $g \in \mathrm{SU}(2)$, so that it satisfies the [above symmetry condition](symmetric_tensor). Therefore, we can represent it as an $ ...
````
```suggestion
where $\vec{S} = (S^x, S^y, S^z)$ are the spin operators. The physical Hilbert space at each site is the three-dimensional spin-1 irrep of $\mathrm{SU}(2)$. Each two-site exchange operator $\vec{S}_i \cdot \vec{S}_j$ in the sum commutes with a global transformation $g \in \mathrm{SU}(2)$, so that it satisfies the [above symmetry condition](symmetric_tensor). Therefore, we can represent it as an $\mathrm{SU}(2)$-symmetric `TensorMap`, as long as we can isolate its reduced matrix elements.
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+    1,1
+\end{smallmatrix} = 1, \quad
+```
+These can then be used to construct the symmetric tensor map representing the exchange interaction:
````
```suggestion
These can then be used to construct the symmetric `TensorMap` representing the exchange interaction:
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+We end this subsection with some comments on the generalization of the above discussion to $\mathrm{SU}(N)$. As foreshadowed above, the irreps of $\mathrm{SU}(N)$ in general have an even more complicated structure. In particular, they can admit so-called *fusion multiplicities*, where the fusion of two irreps can not just have multiple distinct outcomes, but they can even fuse to a given irrep in ...
+```{code-cell} julia
+:tags: [hide-output]
+l = SUNIrrep((2, 1, 0))
````
In principle, you can also use `SU3Irrep("8")`, which I think might be more familiar in the physics community

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+    return (p^2 + q^2 + 3 * p + 3 * q + p * q) / 3
+end
+```
+If we use the adjoint representation of $\mathrm{SU}(3)$ as physical space, the Heisenberg exchage interaction can then be constructed as
````
```suggestion
If we use the adjoint representation of $\mathrm{SU}(3)$ as physical space, the Heisenberg exchange interaction can then be constructed as
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+:scale: 12%
+:name: SUN_fusiontrees
+```
+Using these to directly construct the corresponding symmetric tensor map is much simpler than going through the explicit projection procedure using Clebsch-Gordan coefficients.
````
```suggestion
Using these to directly construct the corresponding symmetric `TensorMap` is much simpler than going through the explicit projection procedure using Clebsch-Gordan coefficients.
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+collect(l ⊗ l)
+```
+However, this particular fusion has multiplicities, since the adjoint irrep can actually fuse to itself in two distinct ways. Indeed, we can use the [`Nsymbol`](https://jutho.github.io/TensorKit.jl/latest/lib/sectors/#TensorKit.Nsymbol) method from TensorKit.jl to inspect the number of times `l` appears in the fusion product `l ⊗ l`, 
+```{code-cell} julia
````
Maybe add the full decomposition:
```{math}
\mathbb{8} \otimes \mathbb{8} = \mathbb{1} \oplus \mathbb{3} \oplus 2 \cdot \mathbb{8} \oplus \mathbb{10} \oplus \mathbb{10^*} \oplus \mathbb{27}
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,793 @@
+:tags: [hide-output]
+V = Vect[FibonacciAnyon](:τ => 1)
+```
+and assigning the following nonzero block value to the two-site fusion trees
````
I think the colon in the figure is a bit strange?

leburgel replied:
It's the same style as every tensor up till then. Is it particularly strange for this one?

lkdvos replied:
I guess I mostly missed the other ones. I just don't think of that relation as a colon, but on the other hand, it does not matter too much so I'm fine either way

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,793 @@
+
+TODO
+
+- Mention something about product symmetries.
```
Hubbard would be cool, but I think we can keep this as a todo for now.

classical stat-mech things with explicit fourier expansion and spherical harmonics are also cool, but maybe also ok as todo for now

lkdvos reviewed (approved):
This is amazing.
I would advice using something either like the VScode extension Reflow Markdown to limit the line length, or just cut lines at the end of a sentence. This also makes git behave better. 
The only remark I would have is distinguishing between what we commonly call big blocks and small blocks a bit more clearly, as this is quite important to work with `TensorMap`s. 
I think tensor map does not look as nice as just `TensorMap`, so mayb a ctrl+f on this would be good as well, as I might have missed a couple

leburgel committed:
Address review comments

leburgel committed:
Reflow markdown

leburgel commented:
I think I addressed all of the comments. I added a little discussion about the 'big' matrix blocks in the actual block-diagonal decomposition in terms of coupled charges, and the 'small' array blocks associated to each fusion tree. It's at the end of section 18.2.1. I hope this clarifies things instead of just making it more confusion. If it's still not clear I can have another go.

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+construct them as `TensorMap`s using an *auxiliary vector space*, based on the following
+intuition. The creation operator $a^+$ violates particle number conservation by mapping the
+occupation number $n$ to $n + 1$. From the point of view of representation theory, this
+process can be thought af the *fusion* of an `U1Irrep(n)` with an `U1Irrep(1)`, naturally
```
```suggestion
process can be thought of as the *fusion* of an `U1Irrep(n)` with an `U1Irrep(1)`, naturally
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+occupation number $n$ to $n + 1$. From the point of view of representation theory, this
+process can be thought af the *fusion* of an `U1Irrep(n)` with an `U1Irrep(1)`, naturally
+giving the fusion product `U1Irrep(n + 1)`. This means we can represent $a^+$ as a
+`TensorMap(..., V ← V ⊗ A)`, where the auxiliary vector space `A` is contains the $+1$ irrep
```
```suggestion
`TensorMap(..., V ← V ⊗ A)`, where the auxiliary vector space `A` contains the $+1$ irrep
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,1240 @@
+:name: bosonops
+```
+
+We can then combine these operators in to get the appropriate Hamiltonian terms,
````
```suggestion
We can then combine these operators to get the appropriate Hamiltonian terms,
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+permuting the physical and auxiliary indices of the corresponding `TensorMap`s. This
+permutation has no effect on the actual array blocks of the tensors due to the
+[bosonic braiding style](https://jutho.github.io/TensorKit.jl/latest/lib/sectors/#TensorKit.BraidingStyle)
+of $\mathrm{U}(1)$ irreps, so the left and right operators can in essence by seen as the
```
```suggestion
of $\mathrm{U}(1)$ irreps, so the left and right operators can in essence be seen as the
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+[bosonic braiding style](https://jutho.github.io/TensorKit.jl/latest/lib/sectors/#TensorKit.BraidingStyle)
+of $\mathrm{U}(1)$ irreps, so the left and right operators can in essence by seen as the
+'same' tensors. This is no longer the case when considering fermionic systems, where
+permuting indices can in fact change the array blocks as we will see next. As consequence,
```
```suggestion
permuting indices can in fact change the array blocks as we will see next. As a consequence,
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+@tensor a⁻a⁺_bis[-1 -2; -3 -4] := a⁻[1 -1; -3] * a⁺[-2; -4 1]
+@tensor N_bis[-1 ; -2] := a⁺[-1; 1 2] * a⁻[2 1; -2]
+
+@test norm(a⁺a⁻_bis - a⁺a⁻) ≈ 0 atol=1e-14
+@test norm(a⁻a⁺_bis - a⁻a⁺) ≈ 0 atol=1e-14
+@test norm(N_bis - N) ≈ 0 atol=1e-14
```
```suggestion
@test a⁺a⁻_bis ≈ a⁺a⁻ atol=1e-14
@test a⁻a⁺_bis ≈ a⁻a⁺ atol=1e-14
@test N_bis ≈ N atol=1e-14
```
I think within the @test macro you can just check for equality like this, and avoid the norms alltogether, alternatively just write out `isapprox` should also work, and might be a little more explicit than comparing norms (should result in same code though)

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+This means that the fusion of two irreps always gives a single irrep as the fusion product,
+and that exchanging two irreps in a tensor product is trivial. In practice, this implies
+that for tensors with these symmetries the fusion trees are completely fixed by the
+uncoupled charges, which uniquely define both the inner line and the coupled charge, and
```
```suggestion
uncoupled charges, which uniquely define both the inner lines and the coupled charge, and
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,1240 @@
+\left\{c_i^-, c_j^+\right\} &= \delta_{ij} .\\
+\end{align*}
+```
+These commutation relations justify the choice of the relative minus sign in the hopping and
````
```suggestion
These relations justify the choice of the relative minus sign in the hopping and
```
either remove commutation or just make it anticommutation?

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+\ket{l_4,m_4}$ to the coupled basis state $\ket{k,n}$ respectively.
+
+The Wigner-Eckart theorem dictates that this structure in terms of Clebsch-Gordan
+coefficients is necessary to ensure that the corresponding is symmetric. It is precisely
```
```suggestion
coefficients is necessary to ensure that the corresponding tensor is symmetric. It is precisely
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+operator in the irrep basis and the Clebsch-Gordan coefficients. In the following we
+demonstrate this explicit procedure for the particular example of $G = \mathrm{SU}(2)$.
+However, it should be noted that for general groups the Clebsch-Gordan coefficients may not
+be as easy to compute (if it is even known how). In addition, the procedure for manually
```
```suggestion
be as easy to compute (in general, no closed formulas exist). In addition, the procedure for manually
```
I think in principle it is known how, at the very least for finite groups and lie groups, in the sense that you can come up with an algorithm that computes them, but not in an analytic way. I don't think this is something we need to go into, but this comment might be better

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+[*quadratic Casimir*](https://en.wikipedia.org/wiki/Representation_theory_of_SU(2)#The_Casimir_element),
+commutes with all generators. By
+[Schur's lemma](https://en.wikipedia.org/wiki/Schur%27s_lemma), it must then act
+proportional to the identity on every irrep, where the corresponding eigenvalue is
```
```suggestion
proportionally to the identity on every irrep, where the corresponding eigenvalue is
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+We end this subsection with some comments on the generalization of the above discussion to
+$\mathrm{SU}(N)$. As foreshadowed above, the irreps of $\mathrm{SU}(N)$ in general have an
+even more complicated structure. In particular, they can admit so-called *fusion
+multiplicities*, where the fusion of two irreps can not just have multiple distinct
```
```suggestion
multiplicities*, where the fusion of two irreps can have not only multiple distinct
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+$\mathrm{SU}(N)$. As foreshadowed above, the irreps of $\mathrm{SU}(N)$ in general have an
+even more complicated structure. In particular, they can admit so-called *fusion
+multiplicities*, where the fusion of two irreps can not just have multiple distinct
+outcomes, but they can even fuse to a given irrep in mutliple inequivalent ways. We can
```
```suggestion
outcomes, but they can even fuse to a given irrep in multiple inequivalent ways. We can
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
````diff
@@ -0,0 +1,1240 @@
+Using SUNRepresentations.jl, we can compute the Casimir as
+```{code-cell} julia
+:tags: [hide-output]
+function casimir(l::SUNIrrep)
````
```suggestion
function casimir(l::SU3Irrep)
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+## Level 5: Anyonic Symmetries and the Golden Chain
+
+While we have focussed exclusively on group-like symmetries in our discussion so far, the
+framework of symmetric actually extends beyond groups to so-called
```
```suggestion
framework of symmetric tensors actually extends beyond groups to so-called
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+These are quite exotic symmetries characterized in terms of
+[the topological data of a unitary fusion category](https://jutho.github.io/TensorKit.jl/stable/man/categories/#ss_topologicalfusion).
+While the precise details of all the terms in these statements fall beyond the scope of this
+tutorial, we can give a simple example of a Hamiltonion model with a categorical symmetry
```
```suggestion
tutorial, we can give a simple example of a Hamiltonian model with a categorical symmetry
```

lkdvos commented on lectures/5-Tutorials/SymmetricTensors.md:
```diff
@@ -0,0 +1,1240 @@
+The Hilbert space of a chain of Fibonacci anyons is not a regular tensor product space, but
+rather a *constrained Hilbert space* where the only allowed basis states are labeled by
+valid Fibonacci fusion configurations. In the golden chain model, we define a
+nearest-neighbor Hamiltonian on this Hilbert space by imposing an energy penalty when any
```
```suggestion
nearest-neighbor Hamiltonian on this Hilbert space by imposing an energy penalty when
```

lkdvos reviewed (approved):
just some minor typos left, otherwise looks ready to go

leburgel committed:
Fix typos

leburgel committed:
One more typo
